pkg/app/app.go

```diff
@@ -310,6 +301,98 @@ func (c *App) parseAuthorizationResponse(r *http.Request) (string, error) {
 	return code[0], nil
 }
 
+func (a *App) DeAuthHandler(w http.ResponseWriter, r *http.Request, user accounts.User) error {
+	if tk, err := a.fetchUserCredentials(user); err != nil || tk == nil {
+		fmt.Fprintln(w, "No credentials found")
+		return err
+	}
+	const pageTemplate = `
+<html><head></head><body>
+<form method="post" action="/deauth">
+<p> This action will remove the authorization to access the build API on your behalf until you authorize it again.</p>
+<p> Would you like to proceed?</p>
+<input type="hidden" name="csrf_token" value="%s"></input>
+<button type="submit" style="background-color:blue;color:white;">Yes</button>
+<button type="button" onclick="document.querySelector('body').innerHTML = 'You may close this page'">No</button>
+</body></html>
+`
+	s := session.Session{
+		OAuth2State: randomHexString(),
+	}
+	if err := a.setOrUpdateSession(w, &s); err != nil {
+		return err
+	}
+	_, err := fmt.Fprintf(w, pageTemplate, s.OAuth2State)
+	return err
+}
+
+func (a *App) RescindAuthorization(w http.ResponseWriter, r *http.Request, user accounts.User) error {
+	r.ParseForm()
+	stateSlice, ok := r.PostForm["csrf_token"]
+	if !ok || len(stateSlice) == 0 {
```

Comment:
should we add unit test to make sure this security check does not get lost in future refactoring?

Reply:
Done